Chatkit/Repositories/JoinedRoomsRepository.swift

```diff
@@ -0,0 +1,148 @@
+import Foundation
+
+/// A repository which exposes a collection of all rooms joined by the user.
+///
+/// Construct an instance of this class using `Chatkit.createJoinedRoomsRepository(...)`
+///
+/// ## What is provided
+///
+/// The repository exposes a state with a set of `rooms` which presents the rooms that the current user
+/// is a member of.
+///
+/// ## Receiving live updates
+///
+/// In order to be notified when the contents of the `rooms` or the `state` of the connection changes,
+/// implement the `JoinedRoomsRepositoryDelegate` protocol and assign
+/// the `JoinedRoomsRepository.delegate` property.
+///
+/// ## Understanding the `state` of the repository
+///
+/// The `state` property describes the state of the live update connection, either
+///   - `.initializing`: awaiting the initial set of data, or
+///   - `.connected`: updates are flowing live, or
+///   - `.degraded`: updates may be delayed due to network problems, or
+///   - `.closed`: the connection is closed, no further updates available.
+public class JoinedRoomsRepository: JoinedRoomsRepositoryProtocol {
+    
+    // MARK: - Types
+    
+    typealias Dependencies = HasTransformer
+    
+    // MARK: - Properties
+    
+    private let buffer: Buffer
+    private let connectivityMonitor: ConnectivityMonitor
+    private let dependencies: Dependencies
+    
+    private var versionedState: VersionedState?
+    private var connectionState: ConnectionState
+    
+    /// The current state of the repository.
+    public private(set) var state: State {
+        didSet {
+            if state != oldValue {
+                self.delegate?.joinedRoomsRepository(self, didUpdateState: state)
```

Comment:
As discussed in person, this potentially needs to be executed on the main thread.  Here the code I suggest...
```suggestion
                if let delegate = self.delegate as? JoinedRoomsViewModel {
                    // We know that the JoinedRoomsViewModel has been coded so that it dispatches to its own
                    // delegate on the main thread.  So we do not need to dispatch to the main thread here.
                    delegate.joinedRoomsRepository(self, didUpdateState: state)
                }
                else {
                    DispatchQueue.main.async {
                        self.delegate?.joinedRoomsRepository(self, didUpdateState: self.state)
                    }
                }
```

Reply:
I've already implemented this locally. I only need to fix unit tests that are now failing due to this change.

Reply:
This is now implemented.